lab_1_keywords_tfidf/main.py

```diff
@@ -4,106 +4,144 @@
+        for punc in string.punctuation:
+            text = text.replace(punc, '')
+        clean_words = text.lower().split()
+        return clean_words
+    return None
 
 
 def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list[str]]:
     """
     Excludes stop words from the token sequence
-
     Parameters:
     tokens (List[str]): Original token sequence
     stop_words (List[str]: Tokens to exclude
-
     Returns:
     List[str]: Token sequence that does not include stop words
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (isinstance(tokens, list) and tokens and all(isinstance(token, str) for token in tokens)
+            and isinstance(stop_words, list)):
+        no_stop_words = [token for token in tokens if token not in stop_words]
+        return no_stop_words
+    return None
 
 
 def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
     """
     Composes a frequency dictionary from the token sequence
-
     Parameters:
     tokens (List[str]): Token sequence to count frequencies for
-
     Returns:
     Dict: {token: number of occurrences in the token sequence} dictionary
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (isinstance(tokens, list) and tokens
+            and all(isinstance(token, str) for token in tokens)):
+        freq_dict = {}
+        for token in tokens:
+            if token not in freq_dict:
+                freq_dict[token] = 1
+            else:
+                freq_dict[token] += 1
+        return freq_dict
+    return None
 
 
 def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[list[str]]:
     """
     Extracts a certain number of most frequent tokens
-
     Parameters:
     frequencies (Dict): A dictionary with tokens and
     its corresponding frequency values
     top (int): Number of token to extract
-
     Returns:
     List[str]: Sequence of specified length
     consisting of tokens with the largest frequency
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (dict_type_check(frequencies, str, (int, float)) and frequencies
+            and isinstance(top, int) and top is not (True or False) and top > 0):
+
+        sorted_words = [token for token, freq in sorted(frequencies.items(), key=lambda token: token[1], reverse=True)]
+        top_words = sorted_words[:top]
+        return top_words
+    return None
 
 
 def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
     """
     Calculates Term Frequency score for each word in a token sequence
     based on the raw frequency
-
     Parameters:
     frequencies (Dict): Raw number of occurrences for each of the tokens
-
     Returns:
     dict: A dictionary with tokens and corresponding term frequency score
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if dict_type_check(frequencies, str, int):
```

Comment:
what if frequencies is empty

Reply:
not sure what i should do with it - should i change this function or just the order of type checks in the other funcs?

Reply:
it is just that in some of the functions you do not check whether the dictionary is not empty, so I commented it here meaning that maybe you would want to include this check here

you can check it anywhere, but it is necessary to check 

in `calculate_tf` it is missing

Reply:
i added if empty check to `calculate_tf` and put all if empty checks before dict content check, hope it works this way